include/nbl/builtin/hlsl/math/morton.hlsl

```diff
@@ -0,0 +1,423 @@
+#ifndef _NBL_BUILTIN_HLSL_MATH_MORTON_INCLUDED_
+#define _NBL_BUILTIN_HLSL_MATH_MORTON_INCLUDED_
+
+#include "nbl/builtin/hlsl/concepts/core.hlsl"
+#include "nbl/builtin/hlsl/bit.hlsl"
+#include "nbl/builtin/hlsl/cpp_compat.hlsl"
+#include "nbl/builtin/hlsl/functional.hlsl"
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace morton
+{
+
+namespace impl
+{
+
+// Valid dimension for a morton code
+template <uint16_t D>
+NBL_BOOL_CONCEPT MortonDimension = 1 < D && D < 5;
+
+template<typename T, uint16_t Dim, uint16_t Bits = 8 * sizeof(T) / Dim>
+struct decode_mask;
+
+template<typename T, uint16_t Dim>
+struct decode_mask<T, Dim, 0> : integral_constant<T, 1> {};
+
+template<typename T, uint16_t Dim, uint16_t Bits>
+struct decode_mask : integral_constant<T, (decode_mask<T, Dim, Bits - 1>::value << Dim) | T(1)> {};
+
+template<typename T, uint16_t Dim, uint16_t Bits = 8 * sizeof(T) / Dim>
+NBL_CONSTEXPR T decode_mask_v = decode_mask<T, Dim, Bits>::value;
+
+} //namespace impl
+
+// Up to D = 4 supported
+#define NBL_HLSL_MORTON_MASKS(U, D) _static_cast<vector<U,D> > (vector< U , 4 >(impl::decode_mask_v< U , D >,\
+                                                    impl::decode_mask_v< U , D > << U (1),\
+                                                    impl::decode_mask_v< U , D > << U (2),\
+                                                    impl::decode_mask_v< U , D > << U (3)\
+                                                   ))
+
+// Making this even slightly less ugly is blocked by https://github.com/microsoft/DirectXShaderCompiler/issues/7006
+// In particular, `Masks` should be a `const static` member field instead of appearing in every method using it
+template<typename I, uint16_t D NBL_PRIMARY_REQUIRES(concepts::IntegralScalar<I> && impl::MortonDimension<D>)
```

Comment:
why not add the `Bits` ? As far as I can see is that `vector<I,D>` is the native/cartesian you convert from/to

but since the encode/decode depends on the number of bits you support, you may as well profit from knowing the limits on the maximum and minumum value of the cartesian

Reply:
tbf, `template<bool Signed, uint16_t D, uint16_t Bits>` would work better, cause you could provide conversion operators and constructors (with enable_if_t) for all `vector<I,D>` where `is_signed_v<I>==Signed` and `sizeof(I) <= Bits`

Reply:
btw in this context `Bits` means bits-per-channel, not total bits like the mask stuff above in the `impl` namespace